docs/path-resolution.rst

```diff
@@ -0,0 +1,454 @@
+
+Remapping allows you to have the compiler replace import path prefixes with something else.
+For example you can set up a remapping so that everything imported from the virtual directory
+``github.com/ethereum/dapp-bin/library`` would actually receive source unit names starting with
+``dapp-bin/library``.
+By setting base path to ``/project`` you could then have the compiler find them in
+``/project/dapp-bin/library``
+
+The remappings can depend on a context, which allows you to configure packages to import,
+e.g. different versions of a library of the same name.
+
+.. warning::
+
+    Information about used remappings is stored in contract metadata so modifying them will result
+    in a slightly different bytecode.
+    This means that if you move your project files to different locations and use remappings to
+    avoid having to adjust the source code, your project will compile but will no longer produce the
+    exact same bytecode as without the remappings.
+
+Import remappings have the form of ``context:prefix=target``.
+All files in or below the ``context`` directory that import a file that starts with ``prefix`` are
+redirected by replacing ``prefix`` with ``target``.
+For example, if you clone ``github.com/ethereum/dapp-bin/`` locally to ``/project/dapp-bin``,
+you can use the following in your source file:
+
+::
+
+    import "github.com/ethereum/dapp-bin/library/iterable_mapping.sol" as it_mapping;
+
+Then run the compiler:
+
+.. code-block:: bash
+
+    solc github.com/ethereum/dapp-bin/=dapp-bin/ --base-path /project source.sol
+
+As a more complex example, suppose you rely on a module that uses an old version of dapp-bin that
+you checked out to ``/project/dapp-bin_old``, then you can run:
+
+.. code-block:: bash
+
+    solc module1:github.com/ethereum/dapp-bin/=dapp-bin/ \
+         module2:github.com/ethereum/dapp-bin/=dapp-bin_old/ \
+         --base-path /project \
+         source.sol
+
+This means that all imports in ``module2`` point to the old version but imports in ``module1``
+point to the new version.
+
+Here are the detailed rules governing the behaviour of remappings:
+
+#. **Remappings only affect the translation between import paths and source unit names.**
+
+   Source unit names added via other means cannot be remapped.
+   For example the paths you specify on the command-line and the ones in ``sources.urls`` in
+   Standard JSON are not affected.
+
+    .. code-block:: bash
+
+        solc /project=/contracts /project/contract.sol # source unit name: /project/contract.sol
+
+#. **Context and prefix must match source unit names, not import paths.**
+
+   - This means that you cannot remap ``./`` or ``../`` directly since they are replaced during
+     the translation to source unit name but you can remap the part of the name they are replaced
+     with:
+
+     .. code-block:: bash
+
+         solc ./=a /project=b /project/contract.sol
```

Comment:
Similar as above. I think it would be also nice to add the source unit name here too, but I'm really not sure from where the data will be loaded. I understand that the source unit name for this file will result in `/project/contract.sol`, but was the content loaded from a file located at `b/contract.sol`?

Reply:
I think adding the source unit name here would be helpful :)

Reply:
It's `/project/contract.sol` because paths given on the command-line are not remapped.

Added a comment to clarify that.